drafts/2019-03-15-integration-testing-with-testcontainers.adoc

```diff
@@ -0,0 +1,324 @@
+}
+----
+
+With these test dependencies, we can create and use a type-safe rest client like this:
+[source,java]
+----
+  @ClassRule
+  public static GenericContainer libertyContainer = new GenericContainer("my-service")
+                    .withExposedPorts(9080)
+                    .waitingFor(Wait.forHttp("/myservice"));
+
+  private static PersonService personSvc; 
+
+  @BeforeClass
+  public static void setupClass() {
+    List<Class<?>> providers = new ArrayList<>();
+    providers.add(MyJsonBProvider.class); // trivial implementation of a JAX-RS MessageBodyReader/Writer
+    String appPath = "http://" + libertyContainer.getContainerIpAddress() + ':' + 
+                                 libertyContainer.getFirstMappedPort() + 
+                                 "/myservice"
+    personSvc = JAXRSClientFactory.create(appPath, PersonService.class, providers);
+  }
+
+  @Test
+  public void testGetPerson() {
+    // invokes HTTP POST on http://<my-service-ip>:<my-service-port>/myservice/
+    // with query params name="Bob" age=24 (as defined on PersonService.createPerson())
+    Long bobId = personSvc.createPerson("Bob", 24);
+    // returns a generated ID
+
+    // invokes HTTP GET http://<my-service-ip>:<my-service-port>/myservice/{bobID}
+    // JSON response is automatically deserlialized into Person object
+    Person bob = personSvc.getPerson(bobId);
+
+    // Now that we have a POJO representing the response, verify data with simple JUnit assertions
+    assertEquals("Bob", bob.name);
+    assertEquals(24, bob.age);
+    assertNotNull(bob.id);
+  }
+----
+
+== Adding in a real database
+
+Chances are that in production we would want our `PersonService` to be persisting information into some sort of database. With Testcontainers, we can easily create a database instance for our app to communicate with. In this case I'll use MongoDB, which does not have a first-class integration with Testcontainers, but is still easy to use as a `GenericContainer`.
+
+To accomplish this, we can start by adding another container for MongoDB and exposing its port:
+[source,java]
+----
+    @ClassRule
+    public static GenericContainer mongodb = new GenericContainer("mongo:3.4")
+                    .withExposedPorts(27017)
+                    .waitingFor(Wait.forListeningPort());
+----
+
+With this setup, the test client can communicate with the Mongo container and the Liberty container. However, the Mongo and Liberty containers cannot communicate with each other, since they are in isolated networks. To fix this problem, we can use a Testcontainer `Network` to establish communication between the two containers.
+
+==== Linking two or more containers
+[source,java]
+----
+    @ClassRule
+    public static Network network = Network.newNetwork();
+
+    @ClassRule
+    public static GenericContainer mongodb = new GenericContainer("mongo:3.4")
+                    .withExposedPorts(27017)
```

Comment:
if you don't plan to access MongoDB from the test's host, there is no need to expose it, everything will be container-to-container here